pypesto/ensemble/ensemble.py

```diff
@@ -367,6 +371,155 @@ def from_sample(
+                        'you want to include more vectors you can consider '
+                        'raising the cutoff value or including parameters '
+                        'from the history with the `from_history` function.')
+
+        x_vectors = np.stack(x_vectors, axis=1)
+        return Ensemble(x_vectors=x_vectors,
+                        x_names=result.problem.x_names,
+                        vector_tags=vector_tags,
+                        lower_bound=result.problem.lb_full,
+                        upper_bound=result.problem.ub_full,
+                        **kwargs)
+
+    @staticmethod
+    def from_history(
+            result: Result,
+            cutoff: float = np.inf,
+            max_size: int = np.inf,
+            max_per_start: int = np.inf,
+            **kwargs,
+    ):
+        """Construct an ensemble from the history of an optimization.
+
+        Parameters
+        ----------
+        result:
+            A pyPESTO result that contains an optimization result
+            with history recorded.
+        cutoff:
+            Exclude parameters from the optimization if the nllh
+            is higher than the `cutoff`.
+        max_size:
+            The maximum size the ensemble should be.
+        max_per_start:
+            The maximum number of vectors to be included from a
+            single optimization start.
+
+        Returns
+        -------
+        The ensemble.
+        """
+        if not result.optimize_result.list[0].history.options['trace_record']:
+            logger.warning('The optimize result has no trace. The Ensemble '
+                           'will automatically be created through '
+                           'from_optimization().')
+            return Ensemble.from_optimization_endpoints(result=result,
+                                                        cutoff=cutoff,
+                                                        max_size=max_size,
+                                                        **kwargs)
+        x_vectors = []
+        vector_tags = []
+        x_names = result.problem.x_names
+        lb = result.problem.lb_full
+        ub = result.problem.ub_full
+
+        # calculate the number of starts whose final nllh is below cutoff
+        n_starts = 0
+        for start in result.optimize_result.list:
+            if start['fval'] > cutoff:
+                break
+            n_starts += 1
+
+        if n_starts*max_per_start > max_size:
+            logger.info(f'The number of starts that can contribute an '
+                        'ensemble vector multiplied with max_per_start '
+                        'is higher than max_size. Thus we will lower '
+                        'max_per_start. If you do not want this to '
+                        'happen consider increasing max_size to '
+                        f'{max_per_start*n_starts} or decrease cutoff.')
+            max_per_start = math.floor(max_size/n_starts)
+
+        for i in range(n_starts):
```

Comment:
Please: i_Start or i_MS or iMS or... not `i`. Never `i`. The reason why not is what just happened a few lines below...